File: VKRServer/VKRServer/Crypto.cs
using System.Numerics;

namespace VKRServer
{
    public class Crypto
    {
        public static int Pow(int G, int a)
        {
            int Result = 1;

            while (a > 0)
            {
                if (a % 2 == 1)
                {
                    Result *= G;
                }

                G *= G;
                a /= 2;
            }

            return Result;
        }

        public static BigInteger PowWithMod(BigInteger G, BigInteger a, BigInteger P)
        {
            BigInteger Result = 1;
            BigInteger Value = G % P;

            while (a > 0)
            {
                if (a % 2 == 1)
                {
                    Result = Result * Value % P;
                }

                Value = Value * Value % P;

                a /= 2;
            }

            return Result;
        }

        public static BigInteger GenNum(int ByteSize)
        {
            BigInteger Num = 1;

            for (int i = 0; i < ByteSize; i++)
            {
                Random Random = new Random();
                byte ez = (byte)Random.Next(0, 256);

                Num = Num << 8;

                if (i == 0)
                {
                    Num = Num >>> 1;
                }

                Num |= ez;
            }

            return Num;
        }
    }
}

File: VKRAndroid/app/src/main/java/com/example/vkr/ui/API/Crypto.java
package com.example.vkr.ui.API;

import java.math.BigInteger;
import java.util.Random;

public class Crypto
{
    public static BigInteger PowWithMod(BigInteger G, BigInteger a, BigInteger P)
    {
        BigInteger result = BigInteger.ONE;
        BigInteger value = G.mod(P);

        while (a.compareTo(BigInteger.ZERO) > 0)
        {
            if (a.mod(BigInteger.valueOf(2)).equals(BigInteger.ONE))
            {
                result = result.multiply(value).mod(P);
            }

            value = value.multiply(value).mod(P);
            a = a.divide(BigInteger.valueOf(2));
        }

        return result;
    }

    public static BigInteger GenNum(int byteSize)
    {
        BigInteger num = BigInteger.ONE;
        Random random = new Random();

        for (int i = 0; i < byteSize; i++)
        {
            byte Z = (byte)random.nextInt(256);

            num = num.shiftLeft(8);

            if (i == 0)
            {
                num = num.shiftRight(1);
            }

            num = num.or(BigInteger.valueOf(Z & 0xFF));
        }

        return num;
    }
}
